File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audiencelist/view/AudienceListPopupDialog.java
package com.trtc.uikit.livekit.component.audiencelist.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.trtc.tuikit.common.livedata.Observer;
import com.trtc.tuikit.common.ui.PopupDialog;
import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.component.audiencelist.store.AudienceListState;
import com.trtc.uikit.livekit.component.audiencelist.view.adapter.AudienceListPanelAdapter;

import java.util.LinkedHashSet;

@SuppressLint("ViewConstructor")
public class AudienceListPopupDialog extends PopupDialog {
    private final Context                                         mContext;
    private       AudienceListPanelAdapter                        mAdapter;
    private final AudienceListState                               mAudienceListState;
    private       ImageView                                       mImageBack;
    private       RecyclerView                                    mRecycleAudienceList;
    private final Observer<LinkedHashSet<TUIRoomDefine.UserInfo>> mAudienceListObserver = this::onAudienceListChange;

    public AudienceListPopupDialog(Context context, AudienceListState audienceListState) {
        super(context);
        mContext = context;
        mAudienceListState = audienceListState;
        initView();
    }

    protected void addObserver() {
        mAudienceListState.audienceList.observe(mAudienceListObserver);
    }

    protected void removeObserver() {
        mAudienceListState.audienceList.removeObserver(mAudienceListObserver);
    }

    protected void initView() {
        View view = LayoutInflater.from(mContext).inflate(R.layout.livekit_layout_audience_list_panel, null);

        bindViewId(view);
        initImageBackView();
        initAudienceListView();
        setView(view);
    }

    @Override
    public void onAttachedToWindow() {
        super.onAttachedToWindow();

        addObserver();
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        removeObserver();
    }

    private void bindViewId(View view) {
        mImageBack = view.findViewById(R.id.iv_back);
        mRecycleAudienceList = view.findViewById(R.id.rv_audience_list);
    }

    private void initImageBackView() {
        mImageBack.setOnClickListener(view -> {
            dismiss();
        });
    }

    private void initAudienceListView() {
        mRecycleAudienceList.setLayoutManager(new LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL,
                false));
        mAdapter = new AudienceListPanelAdapter(mContext, mAudienceListState);
        mRecycleAudienceList.setAdapter(mAdapter);
    }

    private void onAudienceListChange(LinkedHashSet<TUIRoomDefine.UserInfo> userInfo) {
        mAdapter.updateData();
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audiencelist/view/AudienceListPopupDialog.kt
package com.trtc.uikit.livekit.component.audiencelist.view

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.trtc.tuikit.common.ui.PopupDialog
import com.trtc.uikit.livekit.R
import com.trtc.uikit.livekit.component.audiencelist.AudienceListView
import com.trtc.uikit.livekit.component.audiencelist.view.adapter.AudienceListPanelAdapter
import io.trtc.tuikit.atomicxcore.api.LiveAudienceState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

@SuppressLint("ViewConstructor")
class AudienceListPopupDialog(
    private val context: Context,
    private val audienceState: LiveAudienceState
) : PopupDialog(context) {

    private lateinit var imageBack: ImageView
    private lateinit var recycleAudienceList: RecyclerView
    private var adapter: AudienceListPanelAdapter? = null
    private var onUserItemClickListener: AudienceListView.OnUserItemClickListener? = null
    private var audienceListListener: Job? = null

    init {
        val view = LayoutInflater.from(context).inflate(R.layout.audience_list_layout_panel, null)
        bindViewId(view)
        initImageBackView()
        initAudienceListView()
        setView(view)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        audienceListListener = CoroutineScope(Dispatchers.Main).launch {
            audienceState.audienceList.collect { audienceList ->
                adapter?.updateData()
            }
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        audienceListListener?.cancel()
    }

    fun setOnUserItemClickListener(listener: AudienceListView.OnUserItemClickListener?) {
        onUserItemClickListener = listener
        adapter?.setOnItemClickListener(onUserItemClickListener)
    }

    private fun bindViewId(view: View) {
        imageBack = view.findViewById(R.id.iv_back)
        recycleAudienceList = view.findViewById(R.id.rv_audience_list)
    }

    private fun initImageBackView() {
        imageBack.setOnClickListener {
            dismiss()
        }
    }

    private fun initAudienceListView() {
        recycleAudienceList.layoutManager = LinearLayoutManager(
            context, LinearLayoutManager.VERTICAL, false
        )
        adapter = AudienceListPanelAdapter(context, audienceState)
        recycleAudienceList.adapter = adapter
        adapter?.setOnItemClickListener(onUserItemClickListener)
    }
}